source/agora/node/Ledger.d

```diff
@@ -798,8 +798,18 @@ public class Ledger
 
             const CR = this.enroll_man.getCommitmentNonce(K, block.header.height);  // commited R
             if (CR == Point.init)
+            {
+                log.trace("Block#{}: Validator {} (idx: {}) Couldn't find commitment for this validator",
+                          block.header.height, K, idx);
                 return "Block: Couldn't find commitment for this validator";
-            Point R = CR + challenge.toPoint();
+            }
+            const preimage = this.enroll_man.getValidatorPreimageAt(K, Height(block.header.height - 1)); // Preimage at previous block height
+            if (preimage == Hash.init)
+            {
+                log.trace("Block#{}: Validator {} (idx: {}) missing preimage", block.header.height, K, idx);
+                continue;
+            }
+            Point R = CR + Scalar(preimage).toPoint();
```

Comment:
This is not correct. If a validator has signed, then its pre-image needs to be present, and we need to be able to find its commitment. Not being able to find the commitment is a bug in the node (I think), but not finding the pre-image (unless there's a bug in the node) indicates that the block is invalid.

Reply:
I am not sure if we will always have the pre image when we want to validate the block. I'll replace with returning error and see if it still works in system integration.

Reply:
Well a validator is either:
- Listed as missing;
- Not listed as missing, nor signing;
- Listed as signing and not listed as missing;

If a validator is listed as missing, and only listed as missing, its pre-image is not taken into account in the computation of the random seed. For the other two cases, it is known and will be taken into account.

Reply:
Updated and is still 🟢